taichi/backends/vulkan/spirv_ir_builder.cpp

```diff
@@ -0,0 +1,940 @@
+      ib_.begin(spv::OpBranch).add(head_label).commit(&func_);
+      ib_.begin(spv::OpLabel).add(head_label).commit(&func_);
+      ib_.begin(spv::OpLoopMerge).add_seq(merge_label, continue_label, spv::LoopControlMaskNone).commit(&func_);
+      ib_.begin(spv::OpBranch).add(body_label).commit(&func_);
+
+      // body part
+      ib_.begin(spv::OpLabel).add(body_label).commit(&func_);
+      Value tmp0 = load_var(ok, t_int32_);
+      Value tmp1 = new_value(t_bool_, ValueKind::kNormal);
+      ib_.begin(spv::OpIEqual).add_seq(t_bool_, tmp1, tmp0, const_i32_zero_).commit(&func_);
+      ib_.begin(spv::OpBranchConditional).add_seq(tmp1, true_label, merge_label).commit(&func_);
+      ib_.begin(spv::OpLabel).add(true_label).commit(&func_);
+      Value tmp2 = load_var(addr_ptr, t_int32_);
+      store_var(old_val, tmp2);
+      Value tmp3 = load_var(old_val, t_int32_);
+      Value tmp4 = new_value(t_fp32_, ValueKind::kNormal);
+      ib_.begin(spv::OpBitcast).add_seq(t_fp32_, tmp4, tmp3).commit(&func_);
+      Value tmp5 = new_value(t_fp32_, ValueKind::kNormal);
+      ib_.begin(spv::OpFAdd).add_seq(t_fp32_, tmp5, tmp4, data).commit(&func_);
+      Value tmp6 = new_value(t_int32_, ValueKind::kNormal);
+      ib_.begin(spv::OpBitcast).add_seq(t_int32_, tmp6, tmp5).commit(&func_);
+      store_var(new_val, tmp6);
+      Value tmp7 = load_var(old_val, t_int32_);
+      Value tmp8 = load_var(new_val, t_int32_);
+      Value tmp9 = new_value(t_int32_, ValueKind::kNormal);
+      auto const_u32_1 = uint_immediate_number(t_uint32_, 1);
+      auto const_u32_0 = uint_immediate_number(t_uint32_, 0);
+      ib_.begin(spv::OpAtomicCompareExchange).add_seq(t_int32_, tmp9, addr_ptr,
+          const_u32_1,
+          const_u32_0,
+          const_u32_0,
+          tmp8, tmp7).commit(&func_);
+      store_var(cas_val, tmp9);
+      Value tmp10 = load_var(cas_val, t_int32_);
+      Value tmp11 = load_var(old_val, t_int32_);
+      Value tmp12 = new_value(t_bool_, ValueKind::kNormal);
+      ib_.begin(spv::OpIEqual).add_seq(t_bool_, tmp12, tmp10, tmp11).commit(&func_);
+      Value tmp13 = new_value(t_int32_, ValueKind::kNormal);
+      ib_.begin(spv::OpSelect).add_seq(t_int32_, tmp13, tmp12, const_i32_one_, const_i32_zero_).commit(&func_);
+      store_var(ok, tmp13);
+      ib_.begin(spv::OpBranch).add(continue_label).commit(&func_);
+
+      // continue part
+      ib_.begin(spv::OpLabel).add(continue_label).commit(&func_);
+      ib_.begin(spv::OpBranch).add(head_label).commit(&func_);
+
+      // merge part
+      ib_.begin(spv::OpLabel).add(merge_label).commit(&func_);
+      Value tmp14 = load_var(old_val, t_int32_);
+      Value tmp15 = new_value(t_fp32_, ValueKind::kNormal);
+      ib_.begin(spv::OpBitcast).add_seq(t_fp32_, tmp15, tmp14).commit(&func_);
+      ib_.begin(spv::OpReturnValue).add(tmp15).commit(&func_);
+      ib_.begin(spv::OpFunctionEnd).commit(&func_);
+      // function end
+    }
+
+    return float_atomic_add_;
```

Comment:
This is a bit crazy, maybe have a few glsl runtime files that are pre-comiled to SPV? (VulkanSDK has glslc built-in, so this should be possible to do for taichi devs)

Reply:
One reason is SPIR-V is SSA form, we can't determine which SSA values would be compiled by shaderc so may have conflict with the values generated by our ir builder

Reply:
There is an in-development SPIRV Linker in SPIR-V Tools. We can look
into that to link our runtime SPV library to the kernels.

Reply:
In the GL backend, we used to have to "bake" in the buffer into the atomic helper, resulting in a few duplicates. (This is now mitigated by https://github.com/taichi-dev/taichi/pull/2486). One nice thing about this SPIR-V based helper is that we can pass in an Int32 pointer directly, so it only needs to be written once. 

IIRC, I once wrote the atomic helper first in GLSL, then compiled it to SPIR-V and replaced the embedded buffer with an Int32 pointer parameter, which worked just fine. Wonder if we can achieve this automatically?